Mathlib/Data/Fin/Tuple/Embedding.lean

```diff
@@ -0,0 +1,135 @@
+
+* `Fin.Embedding.tail`: the tail of an embedding `x : Fin (n + 1) ↪ α`
+
+* `Fin.Embedding.snoc` : from an embedding `x : Fin n ↪ α` and `a : α`
+such that `a ∉ x.range`, construct an embedding `Fin (n + 1) ↪ α`
+by putting `a` at the end.
+
+* `Fin.Embedding.init`: the init of an embedding `x : Fin (n + 1) ↪ α`
+
+* `Fin.Embedding.append` : merges two embeddings `Fin m ↪ α` and `Fin n ↪ α`
+into an embedding `Fin (m + n) ↪ α` if they have disjoint ranges
+
+-/
+
+open Function.Embedding Fin Set Nat
+
+namespace Fin.Embedding
+
+variable {α : Type*}
+
+/-- Remove the first element from an injective (n + 1)-tuple -/
+def tail {n : ℕ} (x : Fin (n + 1) ↪ α) : Fin n ↪ α :=
+  ⟨Fin.tail x, fun i j h ↦ by rw [← succ_inj, x.inj' h]⟩
+
+theorem coe_tail {n : ℕ} (x : Fin (n + 1) ↪ α) : ↑(tail x) = Fin.tail x := rfl
+
+/-- Adding a new element at the beginning of an injective n-tuple, to get an injective n+1-tuple. -/
+def cons {n : ℕ} (x : Fin n ↪ α) {a : α} (ha : a ∉ range x) : Fin (n + 1) ↪ α :=
+  ⟨Fin.cons a x, cons_injective_iff.mpr ⟨ha, x.inj'⟩⟩
+
+theorem coe_cons {n : ℕ} (x : Fin n ↪ α) {a : α} (ha : a ∉ range x) :
+    ↑(cons x ha) = Fin.cons a x := rfl
+
+theorem tail_cons {n : ℕ} (x : Fin n ↪ α) {a : α} (ha : a ∉ range x) :
+    tail (cons x ha) = x := rfl
+
+/-- Remove the last element from an injective (n + 1)-tuple -/
+def init {n : ℕ} (x : Fin (n + 1) ↪ α) : Fin n ↪ α :=
+  ⟨Fin.init x, fun i j h ↦ by rw [← castSucc_inj, x.inj' h]⟩
+
+/-- Adding a new element at the end of an injective n-tuple, to get an injective n+1-tuple. -/
+def snoc {n : ℕ} (x : Fin n ↪ α) {a : α} (ha : a ∉ range x) : Fin (n + 1) ↪ α :=
+  ⟨Fin.snoc x a, snoc_injective_iff.mpr ⟨x.inj', ha⟩⟩
+
+theorem coe_snoc {n : ℕ} (x : Fin n ↪ α) {a : α} (ha : a ∉ range x) :
+    ↑(snoc x ha) = Fin.snoc x a := rfl
+
+theorem init_snoc {n : ℕ} (x : Fin n ↪ α) {a : α} (ha : a ∉ range x) :
+    init (snoc x ha) = x := by
+  apply coe_injective
+  simp [snoc, init, init_snoc]
+
+theorem snoc_castSucc {n : ℕ} {x : Fin n ↪ α}
+    {a : α} {ha : a ∉ range ⇑x} {i : Fin n} :
+    snoc x ha i.castSucc  = x i := by
+  rw [coe_snoc, Fin.snoc_castSucc]
+
+theorem snoc_last {n : ℕ} {x : Fin n ↪ α}
+    {a : α} {ha : a ∉ range ⇑x} :
+    snoc x ha (last n) = a := by
+  rw [coe_snoc, Fin.snoc_last]
+
+-- Mathlib.Data.Fin.Basic
+theorem _root_.Fin.exists_eq_castAdd_or_exists_eq_natAdd {m n : ℕ} (i : Fin (m + n)) :
+    (∃ i' : Fin m, i = castAdd n i') ∨ ∃ i', i = natAdd m i' := by
+  rcases Nat.lt_or_ge i.val m with hi | hi
+  · exact Or.inl ⟨⟨i.val, hi⟩, by rw [castAdd_mk]⟩
+  · exact Or.inr
+      ⟨⟨i.val - m, by simp [Nat.sub_lt_iff_lt_add' hi, i.prop]⟩,
+        by simp only [natAdd_mk, ← Fin.val_inj, Nat.add_sub_of_le hi]⟩
+
+theorem _root_.Fin.append_injective_iff {m n : ℕ} {x : Fin m → α} {y : Fin n → α} :
+  Function.Injective (Fin.append x y) ↔
+    Function.Injective x ∧ Function.Injective y ∧ Disjoint (range x) (range y) := by
```

Comment:
Thanks for writing this one! Can you move it into the `Basic` file?

Reply:
I did, but that required to add two `import`.